api/types/container/network_test.go

```diff
@@ -0,0 +1,653 @@
+package container
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+)
+
+type TestRanger interface {
+	Range() PortRange
+}
+
+var _ TestRanger = Port{}
+var _ TestRanger = PortRange{}
+
+func TestPort(t *testing.T) {
+	t.Run("Zero Value", func(t *testing.T) {
+		var p Port
+		if !p.IsZero() {
+			t.Errorf("Port{}.IsZero() = false, want true")
+		}
+		if p.IsValid() {
+			t.Errorf("Port{}.IsValid() = true, want false")
+		}
+		if p.String() != "invalid port" {
+			t.Errorf("Port{}.String() = %q, want %q", p.String(), "invalid port")
+		}
+
+		t.Run("Marshal Unmarshal", func(t *testing.T) {
+			var p Port
+			bytes, err := p.MarshalText()
+			if err != nil {
+				t.Errorf("Port{}.MarshalText() error: %v", err)
+			}
+			if len(bytes) != 0 {
+				t.Errorf("Port{}.MarshalText() = %q, want empty string", string(bytes))
+			}
+
+			err = p.UnmarshalText([]byte(""))
+			if err != nil {
+				t.Errorf("Port{}.UnmarshalText(\"\") error: %v", err)
+			}
+			if p != (Port{}) {
+				t.Errorf("Port{}.UnmarshalText(\"\") = %#v, want %#v", p, Port{})
+			}
+		})
+
+		t.Run("JSON Marshal Unmarshal", func(t *testing.T) {
+			var p Port
+			bytes, err := json.Marshal(p)
+			if err != nil {
+				t.Errorf("json.Marshal(Port{}) error: %v", err)
+			}
+			if string(bytes) != `""` {
+				t.Errorf("json.Marshal(Port{}) = %q, want %q", string(bytes), `""`)
+			}
+
+			err = json.Unmarshal([]byte(`""`), &p)
+			if err != nil {
+				t.Errorf("json.Unmarshal(`\"\"`) error: %v", err)
+			}
+			if p != (Port{}) {
+				t.Errorf("json.Unmarshal(`\"\"`) = %#v, want %#v", p, Port{})
+			}
+		})
+	})
+
+	t.Run("PortFrom", func(t *testing.T) {
+		tests := []struct {
+			num   uint16
+			proto PortProto
+		}{
+			{0, TCP},
+			{80, TCP},
+			{8080, TCP},
+			{65535, TCP},
+			{80, UDP},
+			{8080, SCTP},
+		}
+
+		for _, tc := range tests {
+			t.Run(fmt.Sprintf("%d_%s", tc.num, tc.proto), func(t *testing.T) {
+				p, ok := PortFrom(tc.num, tc.proto)
+				if !ok {
+					t.Fatalf("PortFrom(%d, %q) = false, want true", tc.num, tc.proto)
+				}
+				if p.Num() != tc.num {
+					t.Errorf("PortFrom(%d, %q).Num() = %d, want %d", tc.num, tc.proto, p.Num(), tc.num)
+				}
+				if p.Proto() != tc.proto {
+					t.Errorf("PortFrom(%d, %q).Proto() = %q, want %q", tc.num, tc.proto, p.Proto(), tc.proto)
+				}
+			})
+		}
+
+		t.Run("Normalize Protocol", func(t *testing.T) {
+			pr1 := portFrom(1234, "tcp")
+			pr2 := portFrom(1234, "TCP")
+			pr3 := portFrom(1234, "tCp")
+			if pr1 != pr2 || pr2 != pr3 {
+				t.Errorf("PortFrom protocol normalization failed: %q, %q, %q", pr1, pr2, pr3)
+			}
+		})
+
+		negativeTests := []struct {
+			num   uint16
+			proto PortProto
+		}{
+			{0, ""},
+			{80, ""},
+		}
+		for _, tc := range negativeTests {
+			t.Run(fmt.Sprintf("%d_%s", tc.num, tc.proto), func(t *testing.T) {
+				p, ok := PortFrom(tc.num, tc.proto)
+				if ok {
+					t.Fatalf("PortFrom(%d, %q) = true, want false", tc.num, tc.proto)
```

Comment:
```suggestion
					t.Errorf("PortFrom(%d, %q) = true, want false", tc.num, tc.proto)
```
The results of the other checks would still be useful in the failure case. Knowing the properties of the returned Port would help with debugging.